File: src/main/java/com/toast1ng/webtoon/product/domain/WebtoonProduct.java
package com.toast1ng.webtoon.product.domain;

import lombok.Builder;
import lombok.Data;

import java.math.BigDecimal;
import java.time.DayOfWeek;

@Data
public class WebtoonProduct {
    private String id;
    private String title;
    private String authorId;
    private int views;
    private BigDecimal rating;
    private DayOfWeek day;
    private WebtoonStatus status;

    @Builder
    public WebtoonProduct(String id, String title, String authorId, int views, BigDecimal rating, DayOfWeek day, WebtoonStatus status) {
        this.id = id;
        this.title = title;
        this.authorId = authorId;
        this.views = views;
        this.rating = rating;
        this.day = day;
        this.status = status;
    }
}

File: src/main/kotlin/com/toast1ng/webtoon/product/domain/WebtoonProduct.kt
package com.toast1ng.webtoon.product.domain

import java.math.BigDecimal

data class WebtoonProduct(
    val id: Long,
    val title: String,
    val description: String,
    val views: Int,
    val rating: BigDecimal,
    var likes: Int,
    val mainThumbnail: String,
    val subThumbnail: String,
    val day: DayOfWeek,
    val restrictedAge: Int,
    val status: WebtoonStatus,
    val genres: List<Genre>,
    val creators: List<Creator>,
) {
    private val creatorMap = this.creators.associateBy { it.role }

    fun getWriter(): Creator {
        return requireNotNull(this.creatorMap[CreatorRole.WRITER]) { "WRITER is required" }
    }

    fun getIllustrator(): Creator {
        return requireNotNull(this.creatorMap[CreatorRole.ILLUSTRATOR]) { "ILLUSTRATOR is required" }
    }

    fun getOriginalAuthor(): Creator? {
        return this.creatorMap[CreatorRole.ORIGINAL_AUTHOR]
    }

    fun increaseLikeCount() {
        this.likes++
    }

    fun decreaseLikeCount() {
        this.likes--
    }
}